src/Symfony/Component/Config/Definition/BaseNode.php

```diff
@@ -335,8 +400,31 @@ public function getParent()
      */
     final public function finalize($value)
     {
+        if ($value !== $placeholders = self::resolvePlaceholderValue($value)) {
+            foreach ($placeholders as $placeholder) {
+                $this->placeholder = true;
+                try {
+                    $this->finalize($placeholder);
+
+                    return $value;
+                } catch (InvalidConfigurationException $e) {
+                } finally {
+                    $this->placeholder = false;
+                }
+            }
+
+            throw $e;
+        }
+
         $this->validateType($value);
 
+        if (!$this->allowPlaceholders() && $this->placeholder) {
+            $e = new InvalidConfigurationException(sprintf('A dynamic value is not compatible with a "%s" node type at path "%s".', get_class($this), $this->getPath()));
+            $e->setPath($this->getPath());
+
+            throw $e;
+        }
+
         $value = $this->finalizeValue($value);
```

Comment:
In case the value is a placeholder, I think we should preserve the value before the finalization (which will be the placeholder) and return this one instead. Otherwise, transformations done on the placeholder would break it. It should still be called though.

this advice should be challenged though, as it may depend on the usage of the finalization.

Reply:
That's exactly what we do no?

Reply:
See https://github.com/symfony/symfony/pull/23888/files#diff-446b942cb6ccd9c1db2aaff25accc810R415